### PR TITLE
HADOOP-17887. Remove the wrapper class GzipOutputStream

### PR DESCRIPTION
**Pull request by viirya**
As we provide built-in gzip compressor, we can use it in compressor stream. The wrapper `GzipOutputStream` can be removed now.

BTW, I did a microbenchmark by running 10 times of compressing/decompresing random data.

The average time:

After: 12.93s
Before: 13.52s

It is pretty close.

### CODE REVIEW

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java [hunk 1]
```diff
@@ -86,10 +86,6 @@ public int compress(byte[] b, int off, int len) throws IOException {
 
     int compressedBytesWritten = 0;
 
-    if (currentBufLen <= 0) {
-      return compressedBytesWritten;
-    }
-
```
Comment (viirya) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java:92]: Also, found this bug. If we set an empty input to the compress stream, it will cause endless loop.

**Comment by viirya** [comment]
cc @sunchao

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java (hunk 1 above)
Comment (sunchao) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java:92]: can we add a test case for this?

Comment (viirya) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java:92]: Without removing the condition, current test will timeout after removing GzipOutputStream.

Comment (viirya) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java:92]: Do we still need another test for it? We currently have test coverage for it.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 57s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  34m  1s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  22m 45s |  |  trunk passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  compile  |  19m 17s |  |  trunk passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  checkstyle  |   0m 59s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 31s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  1s |  |  trunk passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   1m 34s |  |  trunk passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  spotbugs  |   2m 25s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  18m 33s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 56s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  22m  5s |  |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javac  |  22m  5s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  19m 18s |  |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  javac  |  19m 18s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 57s |  |  hadoop-common-project/hadoop-common: The patch generated 0 new + 6 unchanged - 2 fixed = 6 total (was 8)  |
| +1 :green_heart: |  mvnsite  |   1m 29s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 59s |  |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   1m 36s |  |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  spotbugs  |   2m 34s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  18m 21s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  17m  3s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 49s |  |  The patch does not generate ASF License warnings.  |
|  |   | 189m  2s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3377/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3377 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell |
| uname | Linux bb03b956ac24 4.15.0-153-generic #160-Ubuntu SMP Thu Jul 29 06:54:29 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / a32b475667ac3df98986a1f84b5291e6b7da57bd |
| Default Java | Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3377/1/testReport/ |
| Max. process+thread count | 2016 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3377/1/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java (hunk 1 above)
Comment (sunchao) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java:92]: Which test is failing before removing the condition? so it passes with the wrapper class but fails after?

Also, we can remove the `currentBufLen` variable now since it is no longer used anywhere else.

Comment (viirya) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java:92]: `testGzipCodec` will cause timeout after removing `GzipOutputStream`.

Because of this line:

```java
codecTest(conf, seed, 0, "org.apache.hadoop.io.compress.GzipCodec");
```

It writes an empty input to the compress stream. Due to this `currentBufLen` check, `compress` will return 0 endlessly.

Comment (sunchao) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java:92]: I think it's still good to have a dedicated test for this edge case. We can use `@Test(timeout=<value>)` to check the timeout.

Comment (viirya) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java:92]: i see. let me add one then.

Comment (sunchao) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java:92]: Thanks!

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 56s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  33m 28s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  22m 40s |  |  trunk passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  compile  |  19m 20s |  |  trunk passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  checkstyle  |   1m  0s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 32s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  1s |  |  trunk passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   1m 35s |  |  trunk passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  spotbugs  |   2m 25s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  18m 23s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 56s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  22m 10s |  |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javac  |  22m 10s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  19m 17s |  |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  javac  |  19m 17s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 58s |  |  hadoop-common-project/hadoop-common: The patch generated 0 new + 6 unchanged - 2 fixed = 6 total (was 8)  |
| +1 :green_heart: |  mvnsite  |   1m 28s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m  0s |  |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   1m 36s |  |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  spotbugs  |   2m 40s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  18m 53s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  17m 59s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 46s |  |  The patch does not generate ASF License warnings.  |
|  |   | 189m 49s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3377/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3377 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell |
| uname | Linux 3b7034dde0b3 4.15.0-153-generic #160-Ubuntu SMP Thu Jul 29 06:54:29 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 728a659c3964d81612c917542426a9786984b1c9 |
| Default Java | Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3377/2/testReport/ |
| Max. process+thread count | 1852 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3377/2/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java (hunk 1 above)
Comment (viirya) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java:92]: Added one test for empty input case.

File: hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/io/compress/TestCodec.java [hunk 2]
```diff
@@ -1051,4 +1052,45 @@ public void testCodecPoolAndGzipDecompressor() {
       }
     }
   }
+
+  @Test(timeout=20000)
+  public void testGzipCompressorWithEmptyInput() throws IOException {
```
Comment (viirya) [inline hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/io/compress/TestCodec.java:1057]: In current trunk, this test will cause:

```
org.junit.runners.model.TestTimedOutException: test timed out after 20000 milliseconds         
        at org.apache.hadoop.io.compress.TestCodec.testGzipCompressorWithEmptyInput(TestCodec.java:1076)                                                                                       
```

**Comment by viirya** [comment]
I did a microbenchmark by running 10 times of compressing/decompresing random data.

The average time:

After: 12.93s
Before: 13.52s

It is pretty close.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 58s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  33m 39s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  22m 46s |  |  trunk passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  compile  |  19m 24s |  |  trunk passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  checkstyle  |   1m  1s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 32s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  1s |  |  trunk passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   1m 36s |  |  trunk passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  spotbugs  |   2m 27s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  18m 19s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 56s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  21m 58s |  |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javac  |  21m 58s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  19m 25s |  |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  javac  |  19m 25s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   1m  0s | [/results-checkstyle-hadoop-common-project_hadoop-common.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3377/3/artifact/out/results-checkstyle-hadoop-common-project_hadoop-common.txt) |  hadoop-common-project/hadoop-common: The patch generated 1 new + 61 unchanged - 2 fixed = 62 total (was 63)  |
| +1 :green_heart: |  mvnsite  |   1m 29s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m  0s |  |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   1m 37s |  |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  spotbugs  |   2m 34s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  18m 45s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  17m  6s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 49s |  |  The patch does not generate ASF License warnings.  |
|  |   | 189m 16s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3377/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3377 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell |
| uname | Linux de8f4df59f6d 4.15.0-153-generic #160-Ubuntu SMP Thu Jul 29 06:54:29 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 3442e671682c2717cd7297b4ad009420adde8876 |
| Default Java | Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3377/3/testReport/ |
| Max. process+thread count | 2993 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3377/3/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

**Review by sunchao — APPROVED** [review]
LGTM

**Comment by sunchao** [comment]
Merged to trunk. Thanks @viirya for the contribution!
